### PR TITLE
analyze_pg refactor

### PR DESCRIPTION
**Pull request by BainanXia**
[Pull Request Etiquette doc](https://github.com/Breakthrough-Energy/REISE/wiki/Pull-Request-Etiquette)

### Purpose
This draft PR in its current form serves as a major effort for analyze_pg refactor (see [#323](https://app.zenhub.com/workspaces/renewable-energy-integration-5bfdd1184b5806bc2bf88976/issues/breakthrough-energy/renewableenergyproject/323)). There are two plots we would like to keep and decouple from `analyze_pg` together with all the corresponding support calculations, such as input data preparation and formatting: 

a. time series generation stack plot
b. time series curtailment plot

**One major purpose of this draft PR is to setup an example/standard of plot function implementation in PostREISE so that the code could be useful in many situations, such as informal analysis, formal report generation and front end redistribution, to reduce duplicated efforts.** This could be a long term iterative process.

### What the code is doing
Time series generation stack plot and time series curtailment plot are decoupled from `analyze_pg` and implemented independently in `plot_generation_ts_stack.py` and `plot_curtailment_ts.py.` All support calculations are implemented in different functions and put into proper folders in the current architecture. 

to-dos:
- [x] add tests for support functions once finalized.
- [x] implement time series curtailment plot in a similar way.
- [ ] address comments and potentially conduct major refactor to make it compatible with front end tools. 
- [x] update doc strings.
- [ ] update readme.
- [ ] retire `analyze_pg` officially

### Testing
Integration tests are conducted in the corresponding demo notebooks. The new code generates exactly the same plots as we used in the seams study report. 

### Where to look
Almost everywhere
```
/postreise/analyze/demand.py
/postreise/analyze/helpers.py
/postreise/analyze/generation/summarize.py
/postreise/analyze/generation/capacity_value.py
/postreise/plot/plot_generation_ts_stack.py
/postreise/plot/plot_curtailment_ts.py
/postreise/plot/demo/plot_generation_time_series_stack_demo.ipynb
/postreise/plot/demo/plot_curtailment_time_series_demo.ipynb
```
`plot_generation_ts_stack.py` and `plot_curtailment_ts.py` contain the main body of the plot code.

### Usage Example/Visuals
See `plot_generation_time_series_stack_demo.ipynb` and `plot_curtailment_time_series_demo.ipynb`

### Time estimate
1 hour or more. We would like to dive into details of this PR to come up with a plot function implementation standard, which makes the code useful in multi situations.

### CODE REVIEW

File: postreise/analyze/demand.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def get_demand_time_series(scenario, area, area_type=None):
+    """Get time series demand in certain area of a scenario
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of: *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str area_type: one of: *'loadzone'*, *'state'*,
+        *'state_abbr'*, *'interconnect'*
```
Comment (danielolsen) [inline postreise/analyze/demand.py:15]: Should we simplify this docstring (and make it less likely to drift) by simply saying that these inputs must conform to `area_to_loadzone`? In case we do something like add 'timezone' as another selection criteria to `area_to_loadzone`.

File: postreise/analyze/demand.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def get_net_demand_time_series(scenario, area, area_type=None):
+    """Get time series net demand in certain area of a scenario
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of: *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str area_type: one of: *'loadzone'*, *'state'*,
+        *'state_abbr'*, *'interconnect'*
```
Comment (danielolsen) [inline postreise/analyze/demand.py:33]: Same comment as above.

File: postreise/analyze/generation/capacity_value.py [hunk 3]
```diff
@@ -57,3 +61,39 @@ def calculate_net_load_peak(scenario, resources, hours=100):
...
+    :param str area: one of: *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str area_type: one of: *'loadzone'*, *'state'*,
+        *'state_abbr'*, *'interconnect'*
+    :return: (*float*) -- total storage capacity value
+    """
+    grid = scenario.state.get_grid()
+    storage_id = get_storage_id_in_area(scenario, area, area_type)
+
+    return sum(grid.storage["gen"].loc[storage_id].Pmax.values)
```
Comment (danielolsen) [inline postreise/analyze/generation/capacity_value.py:99]: is this equivalent to `grid.storage["gen"].loc[storage_id].Pmax.sum()`?

File: postreise/analyze/generation/curtailment.py [hunk 4]
```diff
@@ -222,3 +227,40 @@ def summarize_curtailment_by_location(scenario):
...
+                scenario, area, r, area_type=area_type
+            )
+            if r == "wind_offshore":
+                curtailment[r] = (
+                    renewable_profiles["wind"][plant_id].sum(axis=1) - curtailment[r]
+                )
+            else:
+                curtailment[r] = (
+                    renewable_profiles[r][plant_id].sum(axis=1) - curtailment[r]
+                )
```
Comment (danielolsen) [inline postreise/analyze/generation/curtailment.py:263]: Could we simplify this by concatenating all the renewable profiles together? Then we just look up by `plant_id`, no matter the resource.

File: postreise/analyze/generation/summarize.py [hunk 5]
```diff
@@ -129,3 +133,42 @@ def summarize_hist_gen(hist_gen_raw: pd.DataFrame, all_resources: list) -> pd.Da
...
+        resource, index: time stamps, columns: resources
+    """
+    _check_scenario_is_in_analyze_state(scenario)
+    plant_id = get_plant_id_for_resources_in_area(
+        scenario, area, resources, area_type=area_type
+    )
+    pg = scenario.state.get_pg()[plant_id]
+    grid = scenario.state.get_grid()
+
+    return pg.T.groupby(grid.plant["type"]).agg(sum).T
```
Comment (danielolsen) [inline postreise/analyze/generation/summarize.py:157]: This seems to be equivalent to `pg.groupby(grid.plant.type, axis=1).sum()`

File: postreise/analyze/helpers.py [hunk 6]
```diff
@@ -279,3 +280,56 @@ def summarize_plant_to_location(df, grid):
     location_data = df.T.groupby(locations_in_df).sum().T
```
Comment (danielolsen) [inline postreise/analyze/helpers.py:280]: Not part of this PR, but similar to `postreise.analyze.generation.summarize.get_generation_time_series_by_resources`, this can be simplified to `pg.groupby(locations_in_df, axis=1).sum()`.

File: postreise/analyze/helpers.py [hunk 7]
```diff
@@ -279,3 +280,56 @@ def summarize_plant_to_location(df, grid):
...
+        *interconnect*, *'all'*
+    :param str/list resources: one or a list of resources
+    :param str area_type: one of: *'loadzone'*, *'state'*,
+        *'state_abbr'*, *'interconnect'*
+    :return: (*list*) -- list of plant id
+    """
+    if isinstance(resources, str):
+        resource_set = set([resources])
+    else:
+        resource_set = set(resources)
```
Comment (danielolsen) [inline postreise/analyze/helpers.py:300]: `resource_set = set([resources]) if isinstance(resources, str) else set(resources)`?

File: postreise/analyze/generation/summarize.py (hunk 5 above)
Comment (BainanXia) [reply postreise/analyze/generation/summarize.py:157]: Yeah, good catch. I simply copied this line from `analyze_pg`. Will simplify it as you suggested.

File: postreise/analyze/generation/capacity_value.py (hunk 3 above)
Comment (BainanXia) [reply postreise/analyze/generation/capacity_value.py:99]: I believe so.

File: postreise/analyze/helpers.py [hunk 8]
```diff
@@ -279,3 +280,56 @@ def summarize_plant_to_location(df, grid):
...
+
+    storage_id = (
+        grid.storage["gen"]
+        .loc[
+            grid.storage["gen"]["bus_id"]
+            .apply(lambda x: grid.bus.loc[x, "zone_id"])
+            .isin(loadzone_id_set)
+        ]
+        .index.tolist()
+    )
```
Comment (danielolsen) [inline postreise/analyze/helpers.py:333]:
```python
zone_buses = grid.bus.loc[grid.bus.zone_id.isin(loadzone_id_set)].index
storage_gen = grid.storage["gen"]
storage_id = storage_gen.loc[storage_gen.bus_id.isin(zone_buses)].index.tolist()
```

File: postreise/analyze/demand.py (hunk 1 above)
Comment (BainanXia) [reply postreise/analyze/demand.py:15]: The reason I did copy and paste doc strings from `area_to_loadzone` to here instead of simply referring to `area_to_loadzone` is this is a function that a user may call directly and one could simply do `print(get_demand_time_series.__doc__)` to see what parameters are expecting from the function.

Comment (danielolsen) [reply postreise/analyze/demand.py:15]: That's fair. Being explicit is fine, let's just remember this if/when we ever change `area_to_loadzone`!

File: postreise/analyze/helpers.py (hunk 8 above)
Comment (BainanXia) [reply postreise/analyze/helpers.py:333]: Either way, less lines vs less memory usage.

File: postreise/analyze/demand.py (hunk 1 above)
Comment (BainanXia) [reply postreise/analyze/demand.py:15]: 😅

File: postreise/plot/plot_curtailment_ts.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    area of a scenario.
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of: *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str/list resources: one or a list of resources.
+    :param str area_type: one of: *'loadzone'*, *'state'*,
+        *'state_abbr'*, *'interconnect'*
+    :param tuple time_range: (start time stamp, end time stamp). If None, the entire
+        time range is used for the given scenario.
```
Comment (danielolsen) [inline postreise/plot/plot_curtailment_ts.py:57]: I think it would be helpful here to list the allowable type(s) of the timestamps.

File: postreise/plot/plot_curtailment_ts.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+    if t2l:
+        type2label.update(t2l)
+
+    pg = get_generation_time_series_by_resources(
+        scenario, area, resources, area_type=area_type
+    )
+    demand = get_demand_time_series(scenario, area, area_type=area_type)
+    curtailment = get_curtailment_time_series(scenario, area, area_type=area_type)
+
+    if time_zone != "utc":
```
Comment (danielolsen) [inline postreise/plot/plot_curtailment_ts.py:91]: What if the user specifies "UTC" (this is the name of the index of PG currently)? I think the logic of "do we need to actually change the timezone or are they equivalent" is probably better suited to live within `change_time_zone`. In there, we could do something like `if pytz.timezone(time_zone) != pytz.timezone(df.index.name)` (since the timezones of "UTC" and "utc" evaluate to be equal in this way).

File: postreise/plot/plot_curtailment_ts.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    resources = _check_resources_and_format(resources)
+
+    if t2c:
+        type2color.update(t2c)
+    if t2l:
+        type2label.update(t2l)
+
+    pg = get_generation_time_series_by_resources(
+        scenario, area, resources, area_type=area_type
+    )
```
Comment (danielolsen) [inline postreise/plot/plot_curtailment_ts.py:87]: We might want to name this something like `resource_pg`, to be clearer that the columns here are not plant ids (like they are in many other places).

File: postreise/plot/plot_curtailment_ts.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    for r in resources:
+        if r not in pg.columns:
+            raise ValueError(f"{r} is invalid in {area}!")
+        fig = plt.figure(figsize=(20, 10))
+        ax = fig.gca()
+
+        if title:
+            plt.title(title, fontsize=title_fontsize)
+        else:
+            plt.title(f"{area} {r.capitalize()}", fontsize=title_fontsize)
```
Comment (danielolsen) [inline postreise/plot/plot_curtailment_ts.py:129]:
```python
title_text = f"{area} {r.capitalize()}" if title is None else title
plt.title(title_text, fontsize=title_fontsize)
```
?

File: postreise/plot/plot_curtailment_ts.py [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+                label=type2label[cr],
+            )
+            curtailment[cr + "_mean"].plot(
+                ax=ax,
+                ls="--",
+                lw=4,
+                alpha=0.7,
+                color=type2color[cr],
+                label=type2label[cr] + " Mean",
+            )
```
Comment (danielolsen) [inline postreise/plot/plot_curtailment_ts.py:163]: What about something like:
```python
if percentage:
    key1, key2 = (f"{cr}_percentage", f"{cr}_percentage_mean")
else:
   key1, key2 = (cr, f"{cr}_mean")
curtailment[key1].plot(
	ax=ax,
	lw=4,
	alpha=0.7,
	color=type2color[cr],
	label=type2label[cr],
)
curtailment[key2].plot(
	ax=ax,
	ls="--",
	lw=4,
	alpha=0.7,
	color=type2color[cr],
	label=type2label[cr] + " Mean",
)
```
?

File: postreise/plot/plot_curtailment_ts.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+            color=type2color[r],
+            label=f"{type2label[r]} Energy Available",
+        )
+
+        if show_demand:
+            demand.plot(ax=ax_twin, lw=4, alpha=0.7, color="red", label="Demand")
+
+        xt_with_year = list(ax_twin.__dict__["date_axis_info"][0])
+        xt_with_year[-1] = b"%b"
+        ax_twin.__dict__["date_axis_info"][0] = tuple(xt_with_year)
```
Comment (danielolsen) [inline postreise/plot/plot_curtailment_ts.py:179]: A comment on what these lines do would be nice.

File: postreise/plot/plot_curtailment_ts.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+        else:
+            ax.set_ylabel("Curtailment", fontsize=label_fontsize)
+        ax.legend(loc="upper left", prop={"size": legend_fontsize})
+
+        if save:
+            if not filename:
+                filename = f"{area.lower()}_{r}_curtailment"
+            if not filepath:
+                filepath = "./"
+            plt.savefig(f"{filepath + filename}.pdf", bbox_inches="tight", pad_inches=0)
```
Comment (danielolsen) [inline postreise/plot/plot_curtailment_ts.py:201]: This seems like the place to use `os.getcwd` and `os.path.join`.

File: postreise/plot/plot_generation_ts_stack.py [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+    :param str area: one of: *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str/list resources: one or a list of resources. *'solar_curtailment'*,
+        *'wind_curtailment'*, *'wind_offshore_curtailment'* are valid entries together
+        with all available generator types in the area. The order of the resources
+        determines the stack order in the figure.
+    :param str area_type: one of: *'loadzone'*, *'state'*,
+        *'state_abbr'*, *'interconnect'*
+    :param tuple time_range: (start time stamp, end time stamp). If None, the entire
+        time range is used for the given scenario.
```
Comment (danielolsen) [inline postreise/plot/plot_generation_ts_stack.py:68]: Same comment re: specifying allowable timestamp types.

File: postreise/plot/plot_generation_ts_stack.py [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+                capacity_storage_ts, time_range[0], time_range[1]
+            )
+        if time_freq != "H":
+            pg_storage = resample_time_series(pg_storage, time_freq)
+            capacity_storage_ts = resample_time_series(capacity_storage_ts, time_freq)
+
+        pg_stack["storage"] = pg_storage.clip(lower=0)
+        capacity_ts += capacity_storage_ts
+
+        fig, (ax, ax_storage) = plt.subplots(
```
Comment (danielolsen) [inline postreise/plot/plot_generation_ts_stack.py:161]: This is clever!

File: postreise/plot/plot_generation_ts_stack.py [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+
+    if normalize:
+        pg_stack = pg_stack.divide(capacity_ts, axis="index")
+        demand = demand.divide(capacity_ts, axis="index")
+        net_demand = net_demand.divide(capacity_ts, axis="index")
+        ax.set_ylabel("Normalized Generation", fontsize=label_fontsize)
+    else:
+        pg_stack = pg_stack.divide(1000000, axis="index")
+        demand = demand.divide(1000000, axis="index")
+        net_demand = net_demand.divide(1000000, axis="index")
```
Comment (danielolsen) [inline postreise/plot/plot_generation_ts_stack.py:205]: `1000000` -> `1e6`?

File: postreise/plot/plot_generation_ts_stack.py [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+        title = area
+    ax.set_title("%s" % title, fontsize=title_fontsize)
+    ax.grid(color="black", axis="y")
+
+    if "storage" not in resources:
+        # Erase year in xticklabels
+        xt_with_year = list(ax.__dict__["date_axis_info"][0])
+        xt_with_year[-1] = b"%b"
+        ax.__dict__["date_axis_info"][0] = tuple(xt_with_year)
+        ax.set_xlabel("")
```
Comment (danielolsen) [inline postreise/plot/plot_generation_ts_stack.py:228]: Why do we only erase the year when storage is present?

File: postreise/plot/plot_generation_ts_stack.py [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+        loc="upper left",
+        bbox_to_anchor=(1, 1),
+    )
+
+    if save:
+        if not filename:
+            filename = area
+        if not filepath:
+            filepath = "./"
+        plt.savefig(f"{filepath + filename}.pdf", bbox_inches="tight", pad_inches=0)
```
Comment (danielolsen) [inline postreise/plot/plot_generation_ts_stack.py:280]: Same comment re: `os` methods.

**Comment by danielolsen** [comment]
I've gone through all the code so far, the analysis functions are nice and modular and the plotting code is pretty straightforward. I added a few comments where I think things could be a little simpler/clearer, but overall I think this is looking very good.

File: postreise/analyze/generation/summarize.py (hunk 5 above)
Comment (BainanXia) [reply postreise/analyze/generation/summarize.py:157]: It takes me some time to figure it out. In this case, current line will work: `pg.groupby(grid.plant.loc[plant_id, "type"].values, axis=1).sum()`.

File: postreise/plot/plot_curtailment_ts.py (hunk 9 above)
Comment (BainanXia) [reply postreise/plot/plot_curtailment_ts.py:57]: What is the classic doc string format when specifying types of parameters in a tuple, maybe something like this: `:param (pandas.Timestamp/numpy.datetime64/datetime.datetime, pandas.Timestamp/numpy.datetime64/datetime.datetime) time_range:`? @rouille

File: postreise/analyze/generation/summarize.py (hunk 5 above)
Comment (danielolsen) [reply postreise/analyze/generation/summarize.py:157]: Ah, we need the list we pass to the groupby to have the same length as the filtered pg?

File: postreise/plot/plot_curtailment_ts.py (hunk 9 above)
Comment (danielolsen) [reply postreise/plot/plot_curtailment_ts.py:57]: Or `:param tuple time_range: [blah blah blah] where each time stamp is pandas.Timestamp/numpy.datetime64/datetime.datetime`. Does this also work with a str like `"2016-01-01 12:00:00"`?

File: postreise/analyze/generation/capacity_value.py [hunk 21]
```diff
@@ -57,3 +61,39 @@ def calculate_net_load_peak(scenario, resources, hours=100):
...
     top_hours = net_demand.sort_values(ascending=False).head(hours).index
     return resource_generation[top_hours].mean()
+
+
+def get_capacity_by_resources(scenario, area, resources, area_type=None):
+    """Get total capacity value of certain resources in the specific area of a
+    scenario.
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of: *loadzone*, *state*, *state abbreviation*,
```
Comment (rouille) [inline postreise/analyze/generation/capacity_value.py:71]: the second colon is weird. I would remove it, i.e., `one of *'loadzone'*, *'state'*, *'state abbreviation'*, *'interconnect'*, *'all'*

File: postreise/analyze/helpers.py (hunk 8 above)
Comment (rouille) [reply postreise/analyze/helpers.py:333]: I usually prefer less variables

File: postreise/plot/plot_curtailment_ts.py (hunk 9 above)
Comment (rouille) [reply postreise/plot/plot_curtailment_ts.py:57]: Yeah, keep `:param tuple time_range:` and mention the type of the elements in the description as @danielolsen said.

Personally I would not support a `str` ("2016-01-01 12:00:00"), it is a recipe for failure and it is very easy to use `pd.Timestamp`, which by the way accept "2016-01-01 12:00:00":
```
>>> import pandas as pd
>>> pd.Timestamp("2016-01-01 12:00:00")
Timestamp('2016-01-01 12:00:00')
```

File: postreise/analyze/demand.py [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+    """Get time series demand in certain area of a scenario
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of: *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str area_type: one of: *'loadzone'*, *'state'*,
+        *'state_abbr'*, *'interconnect'*
+    :return: (*pandas.Series*) -- time series of total demand, index: time stamps,
+        column demand values
+    """
```
Comment (rouille) [inline postreise/analyze/demand.py:18]: You can use `_check_scenario_is_in_analyze_state` here

File: postreise/analyze/helpers.py (hunk 8 above)
Comment (danielolsen) [reply postreise/analyze/helpers.py:333]: What about something like:
```python
gen = grid.storage["gen"]
storage_id = gen.loc[
	gen["bus_id"].apply(lambda x: grid.bus.loc[x, "zone_id"]).isin(loadzone_id_set)
].index.tolist()
```
?

Getting everything into one expression is nice, but parsing this nine-line expression makes my head hurt a bit. Creating a `gen` variable like this shouldn't have a memory hit since we're just pointing to the same object.

File: postreise/analyze/helpers.py [hunk 23]
```diff
@@ -279,3 +280,56 @@ def summarize_plant_to_location(df, grid):
...
+    else:
+        resource_set = set(resources)
+    grid = scenario.state.get_grid()
+    loadzone_set = area_to_loadzone(grid, area, area_type=area_type)
+    plant_id = grid.plant[
+        (grid.plant["zone_name"].isin(loadzone_set))
+        & (grid.plant["type"].isin(resource_set))
+    ].index.tolist()
+
+    return plant_id
```
[the remark this answers is not in the document]
Comment (rouille) [reply postreise/analyze/helpers.py:308]: I am not sure if we converged here but if you decide to stick with:
```
   plant_id = grid.plant[
        (grid.plant["zone_name"].isin(loadzone_set))
        & (grid.plant["type"].isin(resource_set))
    ].index.tolist()
```
I twould say you have to handle the error in the case `(grid.plant["zone_name"].isin(loadzone_set)) & (grid.plant["type"].isin(resource_set))` is empty.

[the remark this answers is not in the document]
Comment (BainanXia) [reply postreise/analyze/helpers.py:308]: @rouille It will return an empty list in that case and can be properly handled by the rest of the code.

File: postreise/analyze/helpers.py (hunk 8 above)
Comment (BainanXia) [reply postreise/analyze/helpers.py:333]: > What about something like:
> 
> ```python
> gen = grid.storage["gen"]
> storage_id = gen.loc[
> 	gen["bus_id"].apply(lambda x: grid.bus.loc[x, "zone_id"]).isin(loadzone_id_set)
> ].index.tolist()
> ```
> 
> ?
> 
> Getting everything into one expression is nice, but parsing this nine-line expression makes my head hurt a bit. Creating a `gen` variable like this shouldn't have a memory hit since we're just pointing to the same object.

Okay, I will proceed in this way then.

File: postreise/analyze/generation/summarize.py (hunk 5 above)
Comment (BainanXia) [reply postreise/analyze/generation/summarize.py:157]: Exactly, unless we pass function/mapping rule otherwise it won't auto interpreted in the way when `groupby` with default axis.

File: postreise/analyze/demand.py (hunk 22 above)
Comment (BainanXia) [reply postreise/analyze/demand.py:18]: We don't need the scenario to be in `analyze` state when getting demand profile. I will do this check in `get_net_demand_time_series` instead.

Comment (rouille) [reply postreise/analyze/demand.py:18]: No need, it is done in `get_generation_time_series_by_resources`

File: postreise/plot/plot_curtailment_ts.py (hunk 10 above)
Comment (BainanXia) [reply postreise/plot/plot_curtailment_ts.py:91]: Actually, here we are just checking whether the user specified a time zone other than our default value "utc". It won't hurt if we remove this `if` statement and everything still works. In fact, we did `ts.tz_localize("UTC").tz_convert(time_zone) `in `change_time_zone`, and it can handle anything like "UTC", "Utc", "uTc", "uTC", ...

File: postreise/plot/plot_generation_ts_stack.py (hunk 19 above)
Comment (BainanXia) [reply postreise/plot/plot_generation_ts_stack.py:228]: Actually, we did that in both situations, with or without storage. The reason for this code duplication is simply because with storage, the storage axis is the main x-axis, without storage, the original axis is the main x-axis. We only show xticklabels on the main axis.

File: postreise/analyze/generation/capacity_value.py (hunk 3 above)
Comment (BainanXia) [reply postreise/analyze/generation/capacity_value.py]: Done.

File: postreise/analyze/generation/curtailment.py (hunk 4 above)
Comment (BainanXia) [reply postreise/analyze/generation/curtailment.py]: Done.

File: postreise/analyze/generation/summarize.py (hunk 5 above)
Comment (BainanXia) [reply postreise/analyze/generation/summarize.py]: Done.

File: postreise/analyze/helpers.py (hunk 6 above)
Comment (BainanXia) [reply postreise/analyze/helpers.py]: Done.

File: postreise/analyze/helpers.py (hunk 7 above)
Comment (BainanXia) [reply postreise/analyze/helpers.py]: Done.

File: postreise/analyze/helpers.py (hunk 8 above)
Comment (BainanXia) [reply postreise/analyze/helpers.py]: Done.

File: postreise/plot/plot_curtailment_ts.py (hunk 9 above)
Comment (BainanXia) [reply postreise/plot/plot_curtailment_ts.py]: Done.

File: postreise/plot/plot_curtailment_ts.py (hunk 11 above)
Comment (BainanXia) [reply postreise/plot/plot_curtailment_ts.py]: Done.

File: postreise/plot/plot_curtailment_ts.py (hunk 12 above)
Comment (BainanXia) [reply postreise/plot/plot_curtailment_ts.py]: Done.

File: postreise/plot/plot_curtailment_ts.py (hunk 13 above)
Comment (BainanXia) [reply postreise/plot/plot_curtailment_ts.py]: Done.

File: postreise/plot/plot_generation_ts_stack.py (hunk 16 above)
Comment (BainanXia) [reply postreise/plot/plot_generation_ts_stack.py]: Done.

File: postreise/plot/plot_generation_ts_stack.py (hunk 18 above)
Comment (BainanXia) [reply postreise/plot/plot_generation_ts_stack.py]: Done.

File: postreise/plot/plot_generation_ts_stack.py (hunk 20 above)
Comment (BainanXia) [reply postreise/plot/plot_generation_ts_stack.py]: Done.

**Comment by BainanXia** [comment]
@rouille All comments are addressed. Tests for new functions are added. This branch is rebased to the latest `develop` and commit history is cleaned up. Will update Readme after we retire `analyze_pg` completely. Is this a good point to coordinate with the website team and get their suggestions, or this should be handle in a new PR?

File: postreise/analyze/demand.py [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+def get_demand_time_series(scenario, area, area_type=None):
+    """Get time series demand in certain area of a scenario
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str area_type: one of *'loadzone'*, *'state'*, *'state_abbr'*,
+        *'interconnect'*
+    :return: (*pandas.Series*) -- time series of total demand, index: time stamps,
+        column demand values
```
Comment (rouille) [inline postreise/analyze/demand.py:17]: I think there should be a colon `:` after column

File: postreise/analyze/demand.py [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+def get_net_demand_time_series(scenario, area, area_type=None):
+    """Get time series net demand in certain area of a scenario
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str area_type: one of *'loadzone'*, *'state'*, *'state_abbr'*,
+        *'interconnect'*
+    :return: (*pandas.Series*) -- time series of total demand, index: time stamps,
+        column net demand values
```
Comment (rouille) [inline postreise/analyze/demand.py:35]: same here

File: postreise/analyze/generation/summarize.py [hunk 26]
```diff
@@ -129,3 +133,42 @@ def summarize_hist_gen(hist_gen_raw: pd.DataFrame, all_resources: list) -> pd.Da
...
+def get_generation_time_series_by_resources(scenario, area, resources, area_type=None):
+    """Get time series generation for each resource in certain area of a scenario
+
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance
+    :param str area: one of *loadzone*, *state*, *state abbreviation*,
+        *interconnect*, *'all'*
+    :param str/list resources: one or a list of resources
+    :param str area_type: one of *'loadzone'*, *'state'*, *'state_abbr'*,
+        *'interconnect'*
+    :return: (*pandas.DataFrame*) -- data frame of time series generation for each
```
Comment (rouille) [inline postreise/analyze/generation/summarize.py:147]: data frame of time series generation for --> times series of generation for

File: postreise/analyze/demand.py (hunk 24 above)
Comment (BainanXia) [reply postreise/analyze/demand.py]: Done.

File: postreise/analyze/demand.py (hunk 25 above)
Comment (BainanXia) [reply postreise/analyze/demand.py]: Done.

File: postreise/analyze/generation/summarize.py (hunk 26 above)
Comment (BainanXia) [reply postreise/analyze/generation/summarize.py]: Done.

**Review by rouille — APPROVED** [review]
Looks great. Thanks a lot.

File: postreise/analyze/generation/capacity_value.py [hunk 27]
```diff
@@ -57,3 +61,39 @@ def calculate_net_load_peak(scenario, resources, hours=100):
     net_demand = total_demand - resource_generation
     top_hours = net_demand.sort_values(ascending=False).head(hours).index
     return resource_generation[top_hours].mean()
+
+
+def get_capacity_by_resources(scenario, area, resources, area_type=None):
+    """Get total capacity value of certain resources in the specific area of a
```
Comment (danielolsen) [inline postreise/analyze/generation/capacity_value.py:67]: I would rephrase "capacity value" as this might be misinterpreted as something besides the nameplate capacity

File: postreise/analyze/generation/capacity_value.py [hunk 28]
```diff
@@ -57,3 +61,39 @@ def calculate_net_load_peak(scenario, resources, hours=100):
...
+    plant_id = get_plant_id_for_resources_in_area(
+        scenario, area, resources, area_type=area_type
+    )
+    grid = scenario.state.get_grid()
+
+    return grid.plant.loc[plant_id].groupby("type")["Pmax"].sum()
+
+
+def get_storage_capacity(scenario, area, area_type=None):
+    """Get total storage capacity value in the specific area of a scenario.
```
Comment (danielolsen) [inline postreise/analyze/generation/capacity_value.py:87]: Same comment here re "capacity value"

File: postreise/analyze/generation/capacity_value.py (hunk 27 above)
Comment (rouille) [reply postreise/analyze/generation/capacity_value.py:67]: `get_nameplate_capacity` ?

File: postreise/analyze/generation/summarize.py [hunk 29]
```diff
@@ -129,3 +133,42 @@ def summarize_hist_gen(hist_gen_raw: pd.DataFrame, all_resources: list) -> pd.Da
...
+    plant_id = get_plant_id_for_resources_in_area(
+        scenario, area, resources, area_type=area_type
+    )
+    pg = scenario.state.get_pg()[plant_id]
+    grid = scenario.state.get_grid()
+
+    return pg.groupby(grid.plant.loc[plant_id, "type"].values, axis=1).sum()
+
+
+def get_storage_time_series(scenario, area, area_type=None):
```
Comment (danielolsen) [inline postreise/analyze/generation/summarize.py:160]: Should we call this something like `get_storage_pg_time_series`, or have another parameter that specifies whether we want `storage_pg` or `storage_e`? This could easily be in a different PR though, we don't need to hold this one up.

File: postreise/analyze/generation/capacity_value.py (hunk 27 above)
Comment (danielolsen) [reply postreise/analyze/generation/capacity_value.py:67]: I think the function name is fine, I was talking about the docstring. "capacity" by itself doesn't imply the same thing as "capacity value".

**Review by danielolsen — COMMENTED** [review]
This is a big improvement, thanks.

File: postreise/analyze/generation/summarize.py (hunk 29 above)
Comment (BainanXia) [reply postreise/analyze/generation/summarize.py:160]: Done in #215.
